adafruit_connection_manager.py

```diff
@@ -0,0 +1,362 @@
+
+
+def create_fake_ssl_context(
+    socket_pool: SocketpoolModuleType, iface: Optional[InterfaceType] = None
+) -> _FakeSSLContext:
+    """Method to return a fake SSL context for when ssl isn't available to import"""
+    if not iface:
+        # pylint: disable=protected-access
+        iface = socket_pool._the_interface
+    socket_pool.set_interface(iface)
+    return _FakeSSLContext(iface)
+
+
+class ConnectionManager:
+    """Connection manager for sharing sockets."""
+
+    def __init__(
+        self,
+        socket_pool: SocketpoolModuleType,
+    ) -> None:
+        self._socket_pool = socket_pool
+        # Hang onto open sockets so that we can reuse them.
+        self._open_sockets = {}
+        self._socket_free = {}
+
+    def _free_sockets(self) -> None:
+        logger.debug("ConnectionManager.free_sockets()")
+        free_sockets = []
+        for socket, free in self._socket_free.items():
+            if free:
+                key = self._get_key_for_socket(socket)
+                logger.debug(f"  found {key}")
+                free_sockets.append(socket)
+
+        for socket in free_sockets:
+            self.close_socket(socket)
+
+    def _get_key_for_socket(self, socket):
+        try:
+            return next(
+                key for key, value in self._open_sockets.items() if value == socket
+            )
+        except StopIteration:
+            return None
+
+    def close_socket(self, socket: SocketType) -> None:
+        """Close a previously opened socket."""
+        logger.debug("ConnectionManager.close_socket()")
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not managed")
+        key = self._get_key_for_socket(socket)
+        logger.debug(f"  closing {key}")
+        socket.close()
+        del self._socket_free[socket]
+        del self._open_sockets[key]
+
+    def free_socket(self, socket: SocketType) -> None:
+        """Mark a previously opened socket as free so it can be reused if needed."""
+        logger.debug("ConnectionManager.free_socket()")
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not managed")
+        key = self._get_key_for_socket(socket)
+        logger.debug(f"  flagging {key} as free")
+        self._socket_free[socket] = True
+
+    # pylint: disable=too-many-locals,too-many-statements
+    def get_socket(
+        self,
+        host: str,
+        port: int,
+        proto: str,
+        *,
+        timeout: float = 1,
+        is_ssl: bool = False,
+        ssl_context: Optional[SSLContextType] = None,
+        max_retries: int = 5,
+        exception_passthrough: bool = False,
+    ) -> CircuitPythonSocketType:
+        """Get a new socket and connect"""
+        # pylint: disable=too-many-branches
+        logger.debug("ConnectionManager.get_socket()")
+        logger.debug(f"  tracking {len(self._open_sockets)} sockets")
+        key = (host, port, proto)
```

Comment:
I haven't added this change, but would like to entertain the idea of passing in a `session_key: Optinal[str] = None` and using it here:
```
key = (host, port, proto, session_key)
```
This would allow other libraries (like `requests`) to pass in a key that would allow multiple connections to the same host